src/preprocessing/passes/synth_rew_rules.cpp

```diff
@@ -0,0 +1,147 @@
+/*********************                                                        */
+/*! \file synth_rew_rules.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **  Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief synth_rew_rules pass
+ **/
+
+#include "preprocessing/passes/synth_rew_rules.h"
+
+#include "options/base_options.h"
+#include "options/quantifiers_options.h"
+#include "printer/printer.h"
+#include "theory/quantifiers/candidate_rewrite_database.h"
+
+using namespace std;
+
+namespace CVC4 {
+namespace preprocessing {
+namespace passes {
+
+//attribute for whether we have computed rewrite rules for a given term
+struct SynthRrComputedAttributeId {};
+typedef expr::Attribute<SynthRrComputedAttributeId, bool> SynthRrComputedAttribute;
+  
+SynthRewRulesPass::SynthRewRulesPass(PreprocessingPassContext* preprocContext)
+    : PreprocessingPass(preprocContext, "synth-rr"){};
+
+PreprocessingPassResult SynthRewRulesPass::applyInternal(
+    AssertionPipeline* assertionsToPreprocess)
+{
+  Trace("synth-rr-pass") << "Synthesize rewrite rules from assertions..."
+                         << std::endl;
+  std::vector<Node>& assertions = assertionsToPreprocess->ref();
+
+  // compute the variables we will be sampling
+  std::vector<Node> vars;
+  unsigned nsamples = options::sygusSamples();
+
+  Options& nodeManagerOptions = NodeManager::currentNM()->getOptions();
+
+  // attribute to mark processed terms
+  SynthRrComputedAttribute srrca;
```

Comment:
Suggestion: could use a map for consistency with `visited` and since the attribute is not used anywhere else.

Reply:
Actually, it's important that this is a global attribute.  The issue is that in the subcall to the separate SmtEngine, we need to remember that this attribute has been set, which we wouldn't get from a local map. This is to avoid a infinite loop, see issue #1924.

Reply:
Hmm, can you add a comment for that? This is pretty subtle. Also, I should probably submit my patch for having different options for the subsolver at some point...

Reply:
Yeah that would be great.  I'll comment.